guides/unittesting-django-wagtail.md

```diff
@@ -0,0 +1,488 @@
+            self.client.post(reverse('manage-cdn'),
+                             {'url': 'http://fake.gov'})
+            mock_purge.assert_called_with('http://fake.gov')
+    ```
+
+- Logging introspection, to ensure that a message that should be logged does get logged. 
+
+    For example, to mock a module-level logger initialzed with `logger = logging.getLogger(__name__)`, we can patch the the `logging.Logger.info` method and make asserts based on its call arguments:
+
+    ```python
+    @patch('logging.Logger.info')
+    def test_message_gets_logged(self, mock_logger_info):
+        function_that_logs()
+        mock_logger_info.assert_called_with('A message that gets logged')
+    ```
+
+There are other potential uses of Mock, but generally we prefer to test our code operating on real objects as much as as possible rather than mocking.
+
+#### Mocking Requests with Responses
+
+For mocking HTTP calls that are made via [the Requests library](http://docs.python-requests.org/en/master/), we prefer the use of [Responses](https://github.com/getsentry/responses). For example, to test whether an informative banner is displayed to users [the ComplaintLandingView tests](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/legacy/tests/views/test_complaint.py#L67) use responses to provide response data for calls to `requests.get()`:
+
+```python
+from django.test import TestCase
+
+import responses
+
+class ComplaintLandingViewTests(TestCase):
+
+    @responses.activate
+    def test_no_banner_when_data_invalid(self):
+        data_json = {
+            'wrong_key': 5
+        }
+        responses.add(responses.GET, self.test_url, json=data_json)
+        response = ComplaintLandingView.as_view()(self.request)
+        self.assertNotContains(response, 'show-')
+```
+
+#### Mocking boto with moto
+
+When we need to mock AWS services that are called via [boto](https://github.com/boto/boto/) we use the [moto](https://github.com/spulec/moto) library. For example, [to test our S3 utilities](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/v1/tests/test_s3utils.py#L21-L25), we initialize moto in our test case's `setUp` method: 
+
+```python
+class S3UtilsTestCase(TestCase):
+    def setUp(self):
+        mock_s3 = moto.mock_s3()
+        mock_s3.start()
+        self.addCleanup(mock_s3.stop)
+```
+
+From there calls to boto's S3 API will use the moto mock S3.
+
+
+## Running tests
+
+To run Django and Wagtail unit tests we prefer to use [tox](https://tox.readthedocs.io/en/latest/). tox creates and manages virtual environments for running tests against multiple versions of dependencies. 
+
+CFPB has a [sample `tox.ini`](https://github.com/cfpb/development/blob/master/tox.ini) that will test against Django 1.11 and 2.1 and Python 2.7 and 3.6. Additionally, running the tests for [CFPB's cfgov-refresh Django project are documented with that project](https://cfpb.github.io/cfgov-refresh/testing-be/)
```

Comment:
Missing period at the end of this sentence.

Reply:
d8f3f20